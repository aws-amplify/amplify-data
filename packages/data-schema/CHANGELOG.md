# @aws-amplify/data-schema

## 1.2.1

### Patch Changes

- d6114be: fix composite sk in identifier/secondaryIndex

## 1.2.0

### Minor Changes

- f67cb71: adds addToSchema modifier to imported schemas

## 1.1.6

### Patch Changes

- 4c7eb71: perf optimizations, minor builder refactor
- Updated dependencies [4c7eb71]
  - @aws-amplify/data-schema-types@1.0.1

## 1.1.5

### Patch Changes

- 7b044cf: Updated JS Doc strings for relationships, custom queries, mutations, and subscriptions, and custom types

## 1.1.4

### Patch Changes

- 555d0ff: fix identifier - all supported data types; correct SK for list

## 1.1.3

### Patch Changes

- ecc5540: Renaming SQL models should add refersTo directive

## 1.1.2

### Patch Changes

- 7f25eb0: Fixes auth permissions for custom types referenced by custom operations; other minor custom ops fixes

## 1.1.1

### Patch Changes

- 9bf67b7: Use pascal case of model name in graphql operation names

## 1.1.0

### Minor Changes

- 454ab50: feat!: rename relational modifiers for SQL schemas

## 1.0.1

### Patch Changes

- cdfeda5: fix: enforce non-empty a.schema definitions

## 1.0.0

### Major Changes

- 9daf354: first major version

### Patch Changes

- 4fb85b8: chore: add validation to schema process to disallow invalid ref use cases
- Updated dependencies [9daf354]
  - @aws-amplify/data-schema-types@1.0.0

## 0.18.4

### Patch Changes

- a4385c1: fixes return type data prop to always be nullable

## 0.18.3

### Patch Changes

- d473fbb: feat: making enum field or ref of enum field eligible to be PK or SK of 2nd-ary index

## 0.18.2

### Patch Changes

- b3fffcf: feat: rename iam -> identityPool
- Updated dependencies [b3fffcf]
  - @aws-amplify/data-schema-types@0.10.1

## 0.18.1

### Patch Changes

- e06891d: fix: hide .for() for a.query() and a.mutation()

## 0.18.0

### Minor Changes

- d49a92a: restructured ClientSchema properties for extensibility

## 0.17.1

### Patch Changes

- abfbd9d: fix: naming collision on default secondary index queryField

## 0.17.0

### Minor Changes

- e3fd1c7: bump minor in data-schema; missed in previous PR

## 0.16.2

### Patch Changes

- 4f2b22c: update custom operation types to respect optional fields

## 0.16.1

### Patch Changes

- 29d18a8: adds TS types and typesVersions for TS less than 5

## 0.16.0

### Minor Changes

- 2b0f03c: refactor: update authorization builder
- daeafeb: updates builds to include cjs, esm. moves runtime types and behavior into data-schema

### Patch Changes

- d7954e3: fix: timestamp fields are inserted for SQL schema
- Updated dependencies [daeafeb]
  - @aws-amplify/data-schema-types@0.9.0

## 0.15.0

### Minor Changes

- 922c344: feat!: support references based relationships. remove support for manyToMany

### Patch Changes

- Updated dependencies [922c344]
  - @aws-amplify/data-schema-types@0.8.0

## 0.14.14

### Patch Changes

- 4ef3d8d: latest tag is wrong from tag release testing; bumping patch to fix tag
- Updated dependencies [4ef3d8d]
  - @aws-amplify/data-schema-types@0.7.17

## 0.14.13

### Patch Changes

- f106212: fix: only the last function handler can be linked as data source

## 0.14.12

### Patch Changes

- 05915ab: enable model rename for SQL schemas

## 0.14.11

### Patch Changes

- 14a1951: fix: Change interface to move schema combination processing and expose rds config
- Updated dependencies [14a1951]
  - @aws-amplify/data-schema-types@0.7.16

## 0.14.10

### Patch Changes

- 9dfc6a9: feat: add .relationships() modifiers for RDSSchema

## 0.14.9

### Patch Changes

- 62a26bc: fix: Change interface to move schema combination processing and expose rds config
- Updated dependencies [62a26bc]
  - @aws-amplify/data-schema-types@0.7.15

## 0.14.8

### Patch Changes

- 08e47d1: fix: rds custom operations are missing from resolved types

## 0.14.7

### Patch Changes

- 96f3ee4: fix: field with field auth cannot be used as sort key of index

## 0.14.6

### Patch Changes

- 3127247: fix: remove .returns() modifier from custom subscription

## 0.14.5

### Patch Changes

- 0a935ad: feat: Raise an error on transform when combined schemes have conflicting type names

## 0.14.4

### Patch Changes

- 415914e: feat: Update the RDS config to match design feedback

## 0.14.3

### Patch Changes

- ec46bf2: remove implicit graphql fields from graphql schema and model intro schema

## 0.14.2

### Patch Changes

- c5fac1f: feat: ClientSchema conversion of a.combine schemas

## 0.14.1

### Patch Changes

- ffa6675: client types for custom subscriptions
- Updated dependencies [ffa6675]
  - @aws-amplify/data-schema-types@0.7.10

## 0.14.0

### Minor Changes

- 2d85439: refactor!: update .secondaryIndexes builder

## 0.13.18

### Patch Changes

- df1df38: add custom subscriptions

## 0.13.17

### Patch Changes

- 919e01e: feat: RDS ClientSchema without implicit fields / behavior

## 0.13.16

### Patch Changes

- 1804a21: feat: Remove implicit fields for static schema model transformations

## 0.13.15

### Patch Changes

- f33a8e6: remove underscores from generated fn names

## 0.13.14

### Patch Changes

- 0171ad5: handler.function supprts string | defineFunction
- Updated dependencies [0171ad5]
  - @aws-amplify/data-schema-types@0.7.8

## 0.13.13

### Patch Changes

- 0e36d74: fix: delete operation of m:m created implicit model not having correct param

## 0.13.12

### Patch Changes

- ab4bc27: feat: Schema transform logic for inlineSql and sqlReference handlers

## 0.13.11

### Patch Changes

- 14bc3f4: add schema-level lambda access auth rule
- Updated dependencies [14bc3f4]
  - @aws-amplify/data-schema-types@0.7.7

## 0.13.10

### Patch Changes

- bb1c1c6: feat: enable .array() modifier of RefType

## 0.13.9

### Patch Changes

- 1cb4344: feat: Add schema sql handler experience interface

## 0.13.8

### Patch Changes

- 2a56f91: feat: enable more field types and type resolution for CustomType

## 0.13.7

### Patch Changes

- 0cbbdf0: add JS resolver support for handler.custom
- Updated dependencies [0cbbdf0]
  - @aws-amplify/data-schema-types@0.7.6

## 0.13.6

### Patch Changes

- d1e13bc: Add meta types to support custom queries and mutations
- Updated dependencies [d1e13bc]
  - @aws-amplify/data-schema-types@0.7.5

## 0.13.5

### Patch Changes

- ac22ac8: feat: Add custom operation handler types
- Updated dependencies [ac22ac8]
  - @aws-amplify/data-schema-types@0.7.4

## 0.13.4

### Patch Changes

- 78e42c0: custom operations - prevent unrecoverable stack state
- be71ec3: feat: Add internals export to expose configure
- 6e927d1: fix: scope inline enum and custom type names with parent type name
  fix: disallow whitespace in enum values

## 0.13.3

### Patch Changes

- fbcc9dd: type perf improvements
- Updated dependencies [fbcc9dd]
  - @aws-amplify/data-schema-types@0.7.3

## 0.13.2

### Patch Changes

- d52beff: feat: Add addRelationships function to models

## 0.13.1

### Patch Changes

- f2aa9f6: feat: Add model references

## 0.13.0

### Minor Changes

- fdf446c: secondary index client types

### Patch Changes

- Updated dependencies [fdf446c]
  - @aws-amplify/data-schema-types@0.7.0

## 0.12.14

### Patch Changes

- 55e60b7: add secondary index support
- Updated dependencies [55e60b7]
  - @aws-amplify/data-schema-types@0.6.12

## 0.12.13

### Patch Changes

- d5c044a: fix nullable custom type field sel. set
- Updated dependencies [d5c044a]
  - @aws-amplify/data-schema-types@0.6.11

## 0.12.12

### Patch Changes

- 7348b09: depreacte arrayRequired modifier

## 0.12.11

### Patch Changes

- 6181dea: Hide custom ops from top-level client types
- Updated dependencies [6181dea]
  - @aws-amplify/data-schema-types@0.6.9

## 0.12.10

### Patch Changes

- 8370b63: inline custom return types on custom operations
- Updated dependencies [32ea316]
  - @aws-amplify/data-schema-types@0.6.8

## 0.12.9

### Patch Changes

- 8fc3d78: add error on graphql generation for models without authorization rules

## 0.12.8

### Patch Changes

- 83d27dc: fix: removed bad field conflict validation for identifiers

## 0.12.7

### Patch Changes

- 058b3ff: adds protection against conflicting field defs during graphql generation

## 0.12.6

### Patch Changes

- de4d219: fix: Add branding to all schema content types to get errors

## 0.12.5

### Patch Changes

- a74ff3a: removes over-aggressive conflict detection types for model authorization rules

## 0.12.4

### Patch Changes

- cfcf125: fix: Add branding to all field types to get schema errors'
- Updated dependencies [cfcf125]
  - @aws-amplify/data-schema-types@0.6.3

## 0.12.3

### Patch Changes

- a2f4869: docstring corrections for `a.allow` group(s) claim rules
- eb9b549: added missing withClaimIn builder methods on groupDefinedIn, groupsDefinedIn

## 0.12.2

### Patch Changes

- e16e1ca: custom op - add function by name
- Updated dependencies [131ffdd]
- Updated dependencies [504fb2b]
  - @aws-amplify/data-schema-types@0.6.2

## 0.12.1

### Patch Changes

- 87f3717: fix: Remove hasMany relational object mutation field

## 0.12.0

### Minor Changes

- 5df8a96: add custom operations (schema builder & transform)

## 0.11.1

### Patch Changes

- 38c3691: fix: JSON modeling schema issues

## 0.11.0

### Minor Changes

- bcdb110: use approved naming

### Patch Changes

- Updated dependencies [bcdb110]
  - @aws-amplify/data-schema-types@0.4.0

## 0.10.3

### Patch Changes

- c779b83: fix: Remove \_\_date from Authorization export

## 0.10.2

### Patch Changes

- e01c167: updates filter types to include all model fields
- Updated dependencies [e01c167]
  - @aws-amplify/amplify-api-next-types-alpha@0.3.1

## 0.10.1

### Patch Changes

- 6884086: made createdAt, updatedAt fields explicit in generated graphql
- a3e087f: makes owner, group fields explicit in generated graphql models
- 82ce901: made implicit PK field explicit in generated graphql

## 0.10.0

### Minor Changes

- 1f6bfcc: surface relationship FK fields explicitly in model types and generated graphql models

## 0.9.1

### Patch Changes

- 7d0d857: fix: Offer different required modifiers per relationship type

## 0.9.0

### Minor Changes

- df573ad: Added support for "a.allow.custom()" auth rule

## 0.8.0

### Minor Changes

- e7d75b0: custom types and enums support

## 0.7.4

### Patch Changes

- cba8bc9: fix: Mutation inputs should allow incomplete related objects

## 0.7.3

### Patch Changes

- fb0c94c: adds nextToken, limit to list op and related-model loaders
- Updated dependencies [fb0c94c]
  - @aws-amplify/amplify-api-next-types-alpha@0.2.8

## 0.7.2

### Patch Changes

- 2549750: fix relational input field in ModelMeta
- Updated dependencies [2549750]
  - @aws-amplify/amplify-api-next-types-alpha@0.2.7

## 0.7.1

### Patch Changes

- 2d6d480: added filter types for explicit scalar filter fields
- Updated dependencies [2d6d480]
  - @aws-amplify/amplify-api-next-types-alpha@0.2.6

## 0.7.0

### Minor Changes

- ada0afd: rename defineData; remove from public exports

## 0.6.2

### Patch Changes

- 0b72dac: adds docstring

## 0.6.1

### Patch Changes

- 193ffd1: adds authMode, authToken params to model methods and lazy loaders
- Updated dependencies [193ffd1]
  - @aws-amplify/amplify-api-next-types-alpha@0.2.4

## 0.6.0

### Minor Changes

- bf2b31b: update graphql return types to be `{data, nextToken, errors, extensions}`

### Patch Changes

- Updated dependencies [bf2b31b]
  - @aws-amplify/amplify-api-next-types-alpha@0.2.0

## 0.5.0

### Minor Changes

- 227fa88: custom selection set rework

### Patch Changes

- Updated dependencies [227fa88]
  - @aws-amplify/amplify-api-next-types-alpha@0.1.0

## 0.4.1

### Patch Changes

- 623991f: add type-level support for implicit fields
- Updated dependencies [623991f]
  - @aws-amplify/amplify-api-next-types-alpha@0.0.5

## 0.4.0

### Minor Changes

- fb3944a: make model fields optional by default

## 0.3.2

### Patch Changes

- fe1dd56: restructure types
- Updated dependencies [fe1dd56]
- Updated dependencies [fe1dd56]
  - @aws-amplify/amplify-api-next-types-alpha@0.0.3

## 0.3.1

### Patch Changes

- c2db4ec: migrate types to separate types-only package
- Updated dependencies [c2db4ec]
  - @aws-amplify/amplify-api-next-types-alpha@0.0.1

## 0.3.0

### Minor Changes

- 73bfecc: remove default export to fix ESM modules

## 0.2.0

### Minor Changes

- fe93dc2: export ModelSchema type and updated defineData util for Samsara

## 0.1.0

### Minor Changes

- 852f09e: defineData default export conforms to IAmplifyGraphqlDefinition

## 0.0.3

### Patch Changes

- 5146404: fix published package

## 0.0.2

### Patch Changes

- c4675a2: adds initial GHA workflows, changeset, tsc build
