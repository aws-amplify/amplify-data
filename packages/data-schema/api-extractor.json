{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
