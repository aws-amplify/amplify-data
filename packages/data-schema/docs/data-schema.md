<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
