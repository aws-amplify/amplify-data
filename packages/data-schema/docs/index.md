<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md)

## API Reference

## Packages

|  Package | Description |
|  --- | --- |
|  [@aws-amplify/data-schema](./data-schema.md) |  |

