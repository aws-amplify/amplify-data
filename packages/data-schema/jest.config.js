/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  testPathIgnorePatterns: [
    ".test-d.ts", // don't directly execute these files. They'll be tested via their `test.ts` counterparts
    "/node_modules/",
    "utils.ts",
    "/fixtures/",
    "/MappedTypes/",
  ],
  coverageThreshold: {
		global: {
			branches: 72,
			functions: 86,
			lines: 85,
			statements: 85,
		},
	},
};
