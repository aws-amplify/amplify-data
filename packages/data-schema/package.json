{
  "name": "@aws-amplify/data-schema",
  "version": "0.14.31",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-api-next.git"
  },
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.mjs",
  "typings": "./dist/esm/index.d.ts",
  "scripts": {
    "build:esm-cjs": "rollup -c rollup.config.mjs",
    "build": "npm run clean && npm run build:esm-cjs",
    "clean": "rimraf dist tsconfig.tsbuildinfo",
    "docs": "api-extractor run && api-documenter markdown --input-folder temp --output-folder docs",
    "docs:temp": "api-extractor run && api-documenter markdown --input-folder temp --output-folder temp/docs",
    "check-api": "rimraf temp && npm run docs:temp && node ../../scripts/shallow-dir-diff.js docs temp/docs",
    "lint": "eslint .",
    "test": "jest"
  },
  "exports": {
    ".": {
      "types": "./dist/esm/src/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.js",
      "node": "./src/index.ts"
    },
    "./internals": {
      "types": "./dist/esm/src/internals/index.d.ts",
      "import": "./dist/esm/internals/index.mjs",
      "require": "./dist/cjs/internals/index.js",
      "node": "./src/internals/index.ts"
    },
    "./runtime": {
      "types": "./dist/esm/src/runtime/index.d.ts",
      "import": "./dist/esm/runtime/index.mjs",
      "require": "./dist/cjs/runtime/index.js",
      "node": "./src/runtime/index.ts"
    },
    "./package.json": "./package.json"
  },
  "typesVersions": {
    ">=5.0": {
      "internals": [
        "./dist/esm/src/internals/index.d.ts"
      ]
    }
  },
  "dependencies": {
    "@aws-amplify/data-schema-types": "*",
    "@types/aws-lambda": "^8.10.134",
    "rxjs": "^7.8.1"
  },
  "devDependencies": {
    "@changesets/cli": "^2.26.2",
    "@rollup/plugin-typescript": "11.1.5",
    "@tsd/typescript": "^5.1.6",
    "@types/jest": "29.5.4",
    "jest": "^29.7.0",
    "jest-tsd": "^0.2.2",
    "rimraf": "^5.0.5",
    "rollup": "^4.9.6",
    "ts-jest": "^29.1.1",
    "turbo": "^1.10.14",
    "typescript": "^5.1.6"
  },
  "files": [
    "dist/**",
    "internals",
    "runtime",
    "package.json",
    "README.md",
    "CHANGELOG",
    "LICENSE",
    "NOTICE"
  ]
}
