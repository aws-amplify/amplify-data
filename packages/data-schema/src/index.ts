import { schema } from './ModelSchema';
import { model } from './ModelType';
import {
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
} from './ModelField';
import { ref } from './RefType';
import { hasOne, hasMany, belongsTo, manyToMany } from './ModelRelationalField';
import { allow } from './Authorization';
import { customType } from './CustomType';
import { enumType } from './EnumType';
import { query, mutation, subscription } from './CustomOperation';
import { handler } from './Handler';

// We are re-exporting with this pattern to make the JSDoc happy.
// Otherwise the JSDoc wouldn't consistently show up as a customer types
// "a."
export {
  schema,
  model,
  ref,
  customType,
  enumType as enum,
  query,
  mutation,
  subscription,
  hasOne,
  hasMany,
  belongsTo,
  manyToMany,
  allow,
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
  handler,
};
