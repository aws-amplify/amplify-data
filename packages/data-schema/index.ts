import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
