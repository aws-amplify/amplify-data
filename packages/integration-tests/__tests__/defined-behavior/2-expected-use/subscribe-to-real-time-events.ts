import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
  subOptionsAndHeaders,
  useState,
  pause,
} from '../../utils';

const sampleTodo = {
  __typename: 'Todo',
  id: 'some-id',
  content: 'some content',
  description: 'something something',
  owner: 'some-body',
  done: false,
  updatedAt: '2024-03-01T19:05:44.536Z',
  createdAt: '2024-03-01T18:05:44.536Z',
};

describe('Subscribe to real-time events', () => {
  // https://docs.amplify.aws/gen2/build-a-backend/data/subscribe-data/

  // data/resource.ts
  const schema = a.schema({
    Todo: a
      .model({
        content: a.string(),
        description: a.string(),
        done: a.boolean(),
        priority: a.enum(['low', 'medium', 'high']),
      })
      .authorization((allow) => [
        allow.owner(),
        allow.publicApiKey().to(['read']),
      ]),
  });
  type Schema = ClientSchema<typeof schema>;

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('subscription options', async () => {
    // #region mocking
    const { subSpy, subs, generateClient } = mockedGenerateClient([]);

    // simulated amplifyconfiguration.json
    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    // #endregion mocking

    type Todo = Schema['Todo']['type'];
    const client = generateClient<Schema>();
    const [todos, setTodos] = useState<Todo[]>([]);

    // Subscriptions-only `in` operator
    const sub = client.models.Todo.onCreate({
      filter: { priority: { in: ['low', 'medium', 'high'] } },
    }).subscribe({
      next: (data) => {
        setTodos([...todos, data]);
      },
    });

    subs.onCreateTodo.next({
      data: {
        onCreateTodo: {
          ...sampleTodo,
          id: 'some-id-1',
          priority: 'high',
        },
      },
    });

    expect(setTodos.mock.calls).toMatchSnapshot();
    expect(subOptionsAndHeaders(subSpy)).toMatchSnapshot();
  });

  test('subscription options - type level only', async () => {
    const { generateClient } = mockedGenerateClient([]);
    const config = await buildAmplifyConfig(schema);
    Amplify.configure(config);
    const client = generateClient<Schema>();

    // Type-level test only - CRUDL-only filter operator is not allowed for subs
    client.models.Todo.onCreate({
      // @ts-expect-error
      filter: { priority: { attributeExists: true } },
    });
  });
});
