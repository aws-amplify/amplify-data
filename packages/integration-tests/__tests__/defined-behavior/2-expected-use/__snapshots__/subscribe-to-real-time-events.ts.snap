// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribe to real-time events subscription options 1`] = `
[
  [
    [
      {
        "__typename": "Todo",
        "content": "some content",
        "createdAt": "2024-03-01T18:05:44.536Z",
        "description": "something something",
        "done": false,
        "id": "some-id-1",
        "owner": "some-body",
        "priority": "high",
        "updatedAt": "2024-03-01T19:05:44.536Z",
      },
    ],
  ],
]
`;

exports[`Subscribe to real-time events subscription options 2`] = `
[
  [
    {
      "authMode": undefined,
      "query": "subscription ($filter: ModelSubscriptionTodoFilterInput) {
  onCreateTodo(filter: $filter) {
    id
    content
    description
    done
    priority
    createdAt
    updatedAt
    owner
  }
}
",
      "variables": {
        "filter": {
          "priority": {
            "in": [
              "low",
              "medium",
              "high",
            ],
          },
        },
      },
    },
    {},
  ],
]
`;
